README.md
<div align="center">

  <img src="https://on.ahmda.ws/6de47022e721/c" alt="gatsby-remark-better-embed-video" />

  <h1><code>gatsby-remark-better-embed-video</code></h1>

[![GitHub](https://img.shields.io/wordpress/v/akismet.svg?colorA=D14543&colorB=21759B&maxAge=2592000&style=flat&label=WordPress)](https://github.com/ahmadawais/WPGulp/)
[![emoji-log](https://img.shields.io/badge/🚀%20Emoji-Log-gray.svg?colorA=D14543&colorB=21759B&style=flat)](https://github.com/ahmadawais/Emoji-Log/) [![GitHub stars](https://img.shields.io/github/stars/ahmadawais/WPGulp.svg?style=social&label=Stars)](https://github.com/ahmadawais/WPGulp/stargazers) [![GitHub followers](https://img.shields.io/github/followers/ahmadawais.svg?style=social&label=Follow)](https://github.com/ahmadawais?tab=followers)  [![Tweet for help](https://img.shields.io/twitter/follow/mrahmadawais.svg?style=social&label=Tweet%20@MrAhmadAwais)](https://twitter.com/mrahmadawais/)

</div>

<div align="center">
<table width='100%' align="center">
    <tr>
        <td align='left' width='100%' colspan='2'>
            <strong><code>gatsby-remark-better-embed-video</code></strong><br />
            📺 Embed a Youtube Video in your Markdown with more options.
        </td>
    </tr>
    <tr>
        <td>
            A FOSS (Free & Open Source Software) project. Maintained by <a href='https://github.com/ahmadawais'>@AhmadAwais</a>.
        </td>
        <td align='center'>
            <a href='https://twitter.com/MrAhmadAwais/'>
                <img src='https://img.shields.io/badge/→-AHMAD%20AWAIS-gray.svg?colorA=4D2AFF&colorB=2D2B57&style=flat' width='100' />
            </a>
        </td>
    </tr>
</table>
</div>

<br>

Embed a Youtube Video in your Markdown with more options.

Inspired by [gatsby-remark-embed-video](https://github.com/ahmadawais/gatsby-remark-better-embed-video)

## Install

1. Install plugin to your site:

```sh
npm i gatsby-remark-better-embed-video

# OR #

yarn add gatsby-remark-better-embed-video
```

2. Add following to your `gatsby-config.js`:

```js
    plugins: [
      {
        resolve: "gatsby-transformer-remark",
        options: {
          plugins: [
          {
            resolve: "gatsby-remark-better-embed-video",
            options: {
              width: 800,
              ratio: 1.77, // Optional: Defaults to 16/9 = 1.77.
              height: 400, // Optional: Overrides optional.ratio.
              related: false, // Optional: Will remove related videos from the end of an embedded YouTube video.
              noIframeBorder: true, // Optional: Disable insertion of <style> border: 0.
              showInfo: false // Optional: Hides video title and player actions.
            }
          }
          ]
        }
      },
```

Note: if you also rely on `gatsby-remark-responsive-iframe`, you have to define the embed-youtube plugin first:

``` js
plugins: [
  "gatsby-remark-better-embed-video",
  "gatsby-remark-responsive-iframe"
]
```

Note: you can style the videoIframe using `.embedVideoIframe`

1. Restart gatsby.

## Usage

```markdown
# Look at this Video:

`video: https://www.youtube.com/embed/2Xc9gXyf2G4`
`youtube: https://www.youtube.com/watch?v=2Xc9gXyf2G4`
`youtube: 2Xc9gXyf2G4`

`vimeo: https://vimeo.com/5299404`
`vimeo: 5299404`

`videoPress: https://videopress.com/v/kUJmAcSf`
`videoPress: kUJmAcSf`

`twitch: https://player.twitch.tv/?channel=dakotaz`
`twitch: https://player.twitch.tv/?autoplay=false&video=v273436948`
`twitch: 273436948`
`twitchLive: dakotaz`

```

## License

MIT — [Ahmad Awais](https://twitter.com/MrAhmadAwais/)


CHANGELOG.md
# CHANGELOG

All notable changes in this project are documented in this file.

[![Ahmad on Twitter](https://img.shields.io/twitter/follow/mrahmadawais.svg?style=social&label=Follow%20@MrAhmadAwais)](https://twitter.com/mrahmadawais/)

<br>

![hr](https://on.ahmda.ws/t6N5/c)

<br>

### RELEASE: [2.4.0](https://github.com/ahmadawais/gatsby-remark-better-embed-video/compare/2.3.0...2.4.0)

![FIX](https://img.shields.io/badge/-FIX-gray.svg?colorB=ff6347)

> 🐛 Show info AMP [`ad2968e`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/ad2968ee432e40b61e85e598a27c50c2dff4d3bd) <br>

![DOCS](https://img.shields.io/badge/-DOCS-gray.svg?colorB=978CD4)

> 📖 Show Info clean title [`b7d3af8`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/b7d3af864f21e2ac33fbe8c355e2711f2a28ff8b) <br>
> 📖 Center Align [`9bfa25c`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/9bfa25c193358e570a904b3143993c26b67e104f) <br>

<br>

![hr](https://on.ahmda.ws/t6N5/c)

<br>

### RELEASE: [2.3.0](https://github.com/ahmadawais/gatsby-remark-better-embed-video/compare/2.2.0...2.3.0)

![DOCS](https://img.shields.io/badge/-DOCS-gray.svg?colorB=978CD4)

> 📖 Show Info [`5a26ecf`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/5a26ecfb81b7ca0e94baad68b883f697dd3ae27c) <br>

<br>

![hr](https://on.ahmda.ws/t6N5/c)

<br>

### RELEASE: [2.2.0](https://github.com/ahmadawais/gatsby-remark-better-embed-video/compare/2.1.0...2.2.0)

<br>

![hr](https://on.ahmda.ws/t6N5/c)

<br>

### RELEASE: 2.1.0

![NEW](https://img.shields.io/badge/-NEW-gray.svg?colorB=3778FF)

> 📦  Add showInfo option [`5c65418`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/5c6541809352969951cbd44a6c8448a2928ff7d8) <br>
> 📦 Release It [`6b4ed50`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/6b4ed50e16f6766538632380372416038c7416bf) <br>
> 📦 First Commit [`755b044`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/755b044a823e6a8cf6c26ca2fde57f4806cb9b41) <br>

![FIX](https://img.shields.io/badge/-FIX-gray.svg?colorB=ff6347)

> 🐛 Names [`6bf51a8`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/6bf51a8d7eb3756deef5e442d5c4ed460808d6ac) <br>

![DOCS](https://img.shields.io/badge/-DOCS-gray.svg?colorB=978CD4)

> 📖 Show Info [`fa58add`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/fa58adddee2cb3bdd6126875e93125fecdf4d160) <br>
> 📖 header OSS [`84f6574`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/84f6574f5f9a60d007bb6ceddd5156bb882607d1) <br>
> 📖 Improve docs lint [`ce0d13b`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/ce0d13b446ba7cc16864fa7fceca4c7bcdcab45a) <br>
>  📖 Update the changelog [`5a47feb`](https://github.com/ahmadawais/gatsby-remark-better-embed-video/commit/5a47febab3f327bb0f2359271fd7d59b587d331e) <br>

<br>

## OLD CHANGELOG (before fork)

# 1.4.0 (2018-06-15)
* Feature: Twitch Support
* Upgraded Dependencies
* Refactored Package for adding tests

# 1.3.0 (2018-06-02)
* Feature: YouTube Support for additional get parameters + autoconvert parameter "t" to "start" to enable starting the video at a specific starting point.
* Added CSS-Class for easier styling ".embedVideoIframe"
* Added Option "customStyle" to disable auto adding of `<style>`-tag to remove border
* Fix for Issue: "frameborder" attribute obsolute. Remove so we can just use global CSS border:0 on iframe #8
* Fix for Issue: Second VideoPress Link In README Fails #7
* Fix for Issue: Add support for YouTube link with set start time #6

# 1.2.0 (2018-04-20)
* Made the regex matching expression more exact to avoid alternative matchings
# 1.1.2 (2018-04-12)
* Error in package.json
# 1.1.1 (2018-04-12)
* Adjusted default ratio to 1.77
* Refactored function names
# 1.1.0 (2018-04-12)
* Rewrite to typescript

# 1.0.6 (2018-03-30)
* Set default value of related to false.

# 1.0.5 (2018-03-30)
* Fixed Build Error

# 1.0.4 (2018-03-29)
* Added back ratio that was removed by the merge
* Fixed babel build configuration

# 1.0.3 (2018-01-20)
## Features & Bugs
* Added `related` option to remove related videos from the end of an embedded YouTube video.
* Fixed some typo in Readme.md

# 1.0.2 (2018-01-20)
## Bugs
* Removed typos in Readme.md

# 1.0.1 (2018-01-17)
## Bugs
* Fixed links in package.json to point to correct Github Repo


EmbedVideo.ts
/// <reference path="EmbedVideo.d.ts" />

import { URL } from 'url';

const getVideoId = require('get-video-id');

const VideoServices: {
	[key: string]: string;
	YOUTUBE: string;
	VIMEO: string;
	VIDEOPRESS: string;
	TWITCH: string;
	TWITCHLIVE: string;
} = {
	YOUTUBE: 'youtube',
	VIMEO: 'vimeo',
	VIDEOPRESS: 'videopress',
	TWITCH: 'twitch',
	TWITCHLIVE: 'twitchlive'
};

export const getKnownPlatforms = () => {
	return Object.keys(VideoServices).map(val => VideoServices[val] as string);
};

export class EmbedVideo {
	private options: EmbedVideoOptions;

	knownPlatforms = getKnownPlatforms();

	constructor(private type: string, private id: string, options?: EmbedVideoOptions) {
		let defaultOptions = {
			width: 560,
			ratio: 1.77,
			related: false,
			noIframeBorder: false,
			showInfo: true
		};
		this.options = { ...defaultOptions, ...options };

		if (!this.options.height) {
			this.options.height = Math.round(this.options.width / this.options.ratio);
		}
	}

	getHTML() {
		try {
			let videoId = this.readVideoId();
			let url = this.createUrl(videoId.service, videoId.id);
			let iframe = this.createIframe(videoId.service, url);
			return iframe;
		} catch (e) {
			return `<p style="color: red">Error: ${e.message}</p>`;
		}
	}

	private getTwitchId(input: string): VideoId | {} {
		let url;
		try {
			url = new URL(input);
			if (url.origin == 'https://player.twitch.tv') {
				let videoParam = url.searchParams.get('video');
				if (videoParam !== null) {
					return {
						id: videoParam,
						service: VideoServices.TWITCH
					};
				}
				let channelParam = url.searchParams.get('channel');
				if (channelParam !== null) {
					return {
						id: channelParam,
						service: VideoServices.TWITCHLIVE
					};
				}
			}

			if (url.origin == 'https://www.twitch.tv') {
				let pathSplit = url.pathname.split('/');
				if (pathSplit.length >= 2) {
					if (pathSplit[1] === 'videos') {
						if (pathSplit.length > 2) {
							return {
								id: `v${pathSplit[2]}`,
								service: VideoServices.TWITCH
							};
						}
					} else {
						return {
							id: pathSplit[1],
							service: VideoServices.TWITCHLIVE
						};
					}
				}
			}
		} catch (e) {
			return {};
		}

		return {};
	}

	private readVideoId(): VideoId {
		let videoId = getVideoId(this.id);

		if (videoId.id === undefined) {
			videoId = this.getTwitchId(this.id);
		}

		if (videoId.id === undefined) {
			if (this.type === 'video') {
				throw new TypeError('Id could not be processed');
			} else {
				return {
					id: this.id,
					service: this.type.toLowerCase()
				};
			}
		}
		return videoId;
	}

	private createUrl(service: string, videoId: string): string {
		if (service === VideoServices.TWITCH && !videoId.startsWith('v')) {
			videoId = `v${videoId}`;
		}

		const urls: { [index: string]: string } = {
			youtube: `https://www.youtube.com/embed/${videoId}`,
			vimeo: `https://player.vimeo.com/video/${videoId}`,
			videopress: `https://videopress.com/embed/${videoId}`,
			twitch: `https://player.twitch.tv/?autoplay=false&video=${videoId}`,
			twitchlive: `https://player.twitch.tv/?channel=${videoId}`
		};

		const url = new URL(urls[service]);

		if (!url) {
			throw new TypeError('Unknown Video Service');
		}

		if (service === VideoServices.YOUTUBE) {
			if (this.id.startsWith('http')) {
				const originalParams = new URL(this.id);
				originalParams.searchParams.forEach((val, index) => {
					if (index === 'v') {
						// Skip original video Parameter
					} else {
						if (index === 't') {
							let times = val.match(/(\d+)/g);
							if (times) {
								let seconds = times
									.reverse()
									.reduce((total, val, index) => total + parseInt(val) * Math.pow(60, index), 0);
								url.searchParams.set('start', seconds.toString());
							}
						} else {
							url.searchParams.set(index, val);
						}
					}
				});
			}

			if (!this.options.related) {
				url.searchParams.set('rel', '0');
			}

			// If showInfo is false the make the video title etc go away.
			if (!this.options.showInfo) {
				url.searchParams.set('showinfo', '0');
			}
		}

		return url.toString();
	}

	private createIframe(videoPlatform: string, url: string) {
		let iframeNode = `<iframe
              width="${this.options.width}"
              height="${this.options.height}"
              src="${url}"
              class="embedVideoIframe"
              allowfullscreen
            ></iframe>`;
		if (this.options.noIframeBorder) {
			iframeNode += `
      <style>
        .embedVideoIframe {
          border: 0
        }
      </style>`;
		}

		if (videoPlatform === VideoServices.VIDEOPRESS) {
			iframeNode += `<script src="https://videopress.com/videopress-iframe.js"></script>`;
		}

		return iframeNode;
	}
}


EmbedVideo.js
"use strict";
/// <reference path="EmbedVideo.d.ts" />
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var url_1 = require("url");
var getVideoId = require('get-video-id');
var VideoServices = {
    YOUTUBE: 'youtube',
    VIMEO: 'vimeo',
    VIDEOPRESS: 'videopress',
    TWITCH: 'twitch',
    TWITCHLIVE: 'twitchlive'
};
exports.getKnownPlatforms = function () {
    return Object.keys(VideoServices).map(function (val) { return VideoServices[val]; });
};
var EmbedVideo = /** @class */ (function () {
    function EmbedVideo(type, id, options) {
        this.type = type;
        this.id = id;
        this.knownPlatforms = exports.getKnownPlatforms();
        var defaultOptions = {
            width: 560,
            ratio: 1.77,
            related: false,
            noIframeBorder: false,
            showInfo: true
        };
        this.options = __assign({}, defaultOptions, options);
        if (!this.options.height) {
            this.options.height = Math.round(this.options.width / this.options.ratio);
        }
    }
    EmbedVideo.prototype.getHTML = function () {
        try {
            var videoId = this.readVideoId();
            var url = this.createUrl(videoId.service, videoId.id);
            var iframe = this.createIframe(videoId.service, url);
            return iframe;
        }
        catch (e) {
            return "<p style=\"color: red\">Error: " + e.message + "</p>";
        }
    };
    EmbedVideo.prototype.getTwitchId = function (input) {
        var url;
        try {
            url = new url_1.URL(input);
            if (url.origin == 'https://player.twitch.tv') {
                var videoParam = url.searchParams.get('video');
                if (videoParam !== null) {
                    return {
                        id: videoParam,
                        service: VideoServices.TWITCH
                    };
                }
                var channelParam = url.searchParams.get('channel');
                if (channelParam !== null) {
                    return {
                        id: channelParam,
                        service: VideoServices.TWITCHLIVE
                    };
                }
            }
            if (url.origin == 'https://www.twitch.tv') {
                var pathSplit = url.pathname.split('/');
                if (pathSplit.length >= 2) {
                    if (pathSplit[1] === 'videos') {
                        if (pathSplit.length > 2) {
                            return {
                                id: "v" + pathSplit[2],
                                service: VideoServices.TWITCH
                            };
                        }
                    }
                    else {
                        return {
                            id: pathSplit[1],
                            service: VideoServices.TWITCHLIVE
                        };
                    }
                }
            }
        }
        catch (e) {
            return {};
        }
        return {};
    };
    EmbedVideo.prototype.readVideoId = function () {
        var videoId = getVideoId(this.id);
        if (videoId.id === undefined) {
            videoId = this.getTwitchId(this.id);
        }
        if (videoId.id === undefined) {
            if (this.type === 'video') {
                throw new TypeError('Id could not be processed');
            }
            else {
                return {
                    id: this.id,
                    service: this.type.toLowerCase()
                };
            }
        }
        return videoId;
    };
    EmbedVideo.prototype.createUrl = function (service, videoId) {
        if (service === VideoServices.TWITCH && !videoId.startsWith('v')) {
            videoId = "v" + videoId;
        }
        var urls = {
            youtube: "https://www.youtube.com/embed/" + videoId,
            vimeo: "https://player.vimeo.com/video/" + videoId,
            videopress: "https://videopress.com/embed/" + videoId,
            twitch: "https://player.twitch.tv/?autoplay=false&video=" + videoId,
            twitchlive: "https://player.twitch.tv/?channel=" + videoId
        };
        var url = new url_1.URL(urls[service]);
        if (!url) {
            throw new TypeError('Unknown Video Service');
        }
        if (service === VideoServices.YOUTUBE) {
            if (this.id.startsWith('http')) {
                var originalParams = new url_1.URL(this.id);
                originalParams.searchParams.forEach(function (val, index) {
                    if (index === 'v') {
                        // Skip original video Parameter
                    }
                    else {
                        if (index === 't') {
                            var times = val.match(/(\d+)/g);
                            if (times) {
                                var seconds = times
                                    .reverse()
                                    .reduce(function (total, val, index) { return total + parseInt(val) * Math.pow(60, index); }, 0);
                                url.searchParams.set('start', seconds.toString());
                            }
                        }
                        else {
                            url.searchParams.set(index, val);
                        }
                    }
                });
            }
            if (!this.options.related) {
                url.searchParams.set('rel', '0');
            }
            // If showInfo is false the make the video title etc go away.
            if (!this.options.showInfo) {
                url.searchParams.set('amp;showinfo', '0');
            }
        }
        return url.toString();
    };
    EmbedVideo.prototype.createIframe = function (videoPlatform, url) {
        var iframeNode = "<iframe\n              width=\"" + this.options.width + "\"\n              height=\"" + this.options.height + "\"\n              src=\"" + url + "\"\n              class=\"embedVideoIframe\"\n              allowfullscreen\n            ></iframe>";
        if (this.options.noIframeBorder) {
            iframeNode += "\n      <style>\n        .embedVideoIframe {\n          border: 0\n        }\n      </style>";
        }
        if (videoPlatform === VideoServices.VIDEOPRESS) {
            iframeNode += "<script src=\"https://videopress.com/videopress-iframe.js\"></script>";
        }
        return iframeNode;
    };
    return EmbedVideo;
}());
exports.EmbedVideo = EmbedVideo;
